jobs: optimistically update progress, status, and info storage writes

msbutler opened the pull request:
This patch converts the delete-then-insert strategy used to the update progress, info, status of a job to a single update. If the job_id does not exist in the respictive table, then an insert is used.

Epic: none

Release note: none

msbutler committed:
jobs: optimistically update progress, storage, and status writes

This patch converts the delete-then-insert strategy used in these job tables
to a single sql update. If the job_id (+info_key) does not exist in the  table,
then an insert is used. Note that if the update  modifies a row, it implies the
mvcc layer will write a tombstone over the previous key and write a new key, as
the update is writing new primary key (with a new timestamp).

This patch also adds error handling to defend against the update failing with a
uniqueness violation, which could happen if rows were manually inserted into
the tables with the same job_id (and info key).

Epic: none

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141682)
<!-- Reviewable:end -->

stevendanna commented on pkg/jobs/job_info_storage.go:
```diff
@@ -114,12 +107,33 @@ func (i ProgressStorage) Set(
+	// INSERT, but that would have always required two separate SQL operations.
+	rowsUpdated, err := txn.ExecEx(
+		ctx, "write-job-progress-update", txn.KV(), sessiondata.NodeUserSessionDataOverride,
+		`UPDATE system.job_progress SET (written, fraction, resolved) = (now(), $2, $3) WHERE job_id = $1`,
 		i, frac, ts,
-	); err != nil {
-		return err
+	)
```
Can we use an `UPSERT` here? Since this table doesn't have a secondary index and you are setting all columns, I think it should end up with a single 1PC Put.

dt commented:
I wonder if we should add an param or additional method to the API for use during creation when you don't expect a row to exist to skip the (overly) optimistic round-trip before getting to the fallback? I'm just thinking this makes _updates_ cheaper but creations more expensive, but creations are maybe where we're actually most sensitive to cost/time since those happen on the planner's txn, vs a progress update which happens in the background.

msbutler commented:
> I wonder if we should add an param or additional method to the API for use during creation when you don't expect a row to exist to skip the (overly) optimistic round-trip before getting to the fallback?

I like that idea, though it seems this would only apply to writes to job_info, the only auxiliary table touched during job creation. i'll add a follow up pr.

msbutler committed:
jobs: add writeFirstLegacyPayload and writeFirstLegacyProgress internal apis

This patch adds new job info api calls which callers in the jobs package can
use to write the first job info keys. These 2 calls reduce the number of sql
statements executed on the first info key write from 2 to 1.

Epic: none

Release note: none

dt commented:
Can you expand a little on the "It is the caller's responsibility to ensure that the info record does not already exist." choice? How bad is it if you get this wrong (e.g. your function is in a retry loop, which isn't unheard of around job-y things)? What are we getting in return for this limitation/risk?

I was picturing the "first" hint just skipping the optimistic update, and proceeding to the classic path of delete-then-write, which AFAIK wouldn't require this disclaimer, but I'm guessing you went with this approach to save the stmt/scan for the noop delete. I'm not saying that's a bad trade, but curious if you want to expand on the reasoning around it.

msbutler commented:
> Can you expand a little on the "It is the caller's responsibility to ensure that the info record does not already exist." choice? How bad is it if you get this wrong (e.g. your function is in a retry loop, which isn't unheard of around job-y things)? What are we getting in return for this limitation/risk?

Yeah the main benefit is that now job creation only requires a single sql write to job info. in terms of managing the risk you point out: currently the only callers of `WriteFirstLegacy*` are public job apis: `CreateJobWithTxn`, `CreateAdoptableJobWithTxn` and `CreateJobsWithTxn`. I think we can assume that these three funcs will never pass a job ID that already exists in job_info. To prevent/minimize misuse of `WriteFirstLegacy*`, we could make these private funcs, so only funcs in the jobs package can use them. Would privatizing these minimize risk to your satisfaction?

msbutler commented:
TFTR!

bors r=dt

msbutler commented:
bors r+
